### PR TITLE
update Lean version and replace `std` to `batteries`

### PR DESCRIPTION
**Pull request by Seasawher**
This book is very out of date. 

* Std is gone and Batteries must be used.
* Due to Lean updates, some of the code examples do not work as intended.

This PR is the first step in an attempt to update this book.

### CODE REVIEW

File: lakefile.lean [hunk 1]
```diff
@@ -12,7 +12,7 @@ lean_lib «lean4-metaprogramming-book» where
 require mdgen from git
   "https://github.com/Seasawher/mdgen" @ "v1.3.0"
 
-require std from git "https://github.com/leanprover/std4" @ "v4.7.0"
+require "leanprover-community" / "batteries" @ git "main"
```
Comment (Julian) [inline lakefile.lean:15]: Shouldn't we use the corresponding tag -- i.e. https://github.com/leanprover-community/batteries/releases/tag/v4.15.0-rc1 -- so this doesn't break when Batteries updates its Lean version?

File: lean/main/09_tactics.lean [hunk 2]
```diff
@@ -386,7 +386,7 @@ elab "custom_assump_2" : tactic =>
         then return Option.some declExpr
         else return Option.none
     match option_matching_expr with
-    | some e => Lean.Elab.Tactic.closeMainGoal e
+    | some e => Lean.Elab.Tactic.closeMainGoal `custom_assump_2 e
```
Comment (Julian) [inline lean/main/09_tactics.lean:389]: I'm only scanning the PR so apologies for not trying to figure this out myself, but can you say something about what's happening here with the extra argument?

**Comment by Julian** [comment]
I think this looks mostly good to me, just left 2 minor comments. Appreciate the PR!

File: lakefile.lean (hunk 1 above)
Comment (Seasawher) [reply lakefile.lean:15]: Batteries updates will not break the code in this book until a `lake update` is manually performed.

This is not a problem, as any updates to Batteries should eventually be followed.

File: lean/main/09_tactics.lean (hunk 2 above)
Comment (Seasawher) [reply lean/main/09_tactics.lean:389]: The `Lean.Elab.Tactic.closeMainGoal` argument appears to have been increased. It now accepts the tactic name for use in error messages.

**Comment by Julian** [comment]
Got it, ok thanks again!

**Pull request merged**